ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -433,6 +404,36 @@ KEXT_STATIC int HandleVnodeOperation(
         }
     }
     
+    if (isDeleteAction)
+    {
+        // Allow any user to delete individual files, as this generally doesn't cause nested kauth callbacks.
+        ProviderCallbackPolicy callbackPolicy = isDirectory ? CallbackPolicy_UserInitiatedOnly : CallbackPolicy_AllowAny;
+        if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, callbackPolicy, &root, &vnodeFsidInode, &kauthResult, kauthError))
```

Comment:
It looks like we'll have already looked up the root by now for deletes, can we remove this call to `TryGetVirtualizationRoot`?

Reply:
I can, the old way also did 2 looksup, but it's fair to say that was never a hard requirement.

Reply:
Yeah I noticed that too.  I could not think of a reason that was required though, maybe it was just an oversight?

Reply:
I *think* and it would be great to get confirmation on this, since the virtualization roots aren't on lock they could change between calls.  That said nothing is under lock so it really doesn't guarantee you anything just that you checked again a few lines down.

I think it was probably just a code copy oversight.  It's also easier now to tell we have the virtualization root based on where the code is now.

Reply:
There actually is a notable change worth calling out.  For the Predelete we were looking up the root with "CallbackPolicy_AllowAny". 

However the following check for Hydrate would have been called with "CallbackPolicy_UserInitiatedOnly"

@pmj can you elaborate on the desired behavior here?  For Deletes I would expect both the Hydrate and PreDelete to be processed (or neither).

Reply:
Regarding calling `TryGetVirtualizationRoot` multiple times - sorting this out is *very* high on my list of things to do as part of #900, I'm hoping we can do that next after #182 and #526 are dealt with.

Regarding predelete vs hydrate policy: for directories it looks like it's already consistent. For files you're right, it doesn't really make sense to have a discrepancy in policy here. Note that from #1193 on we have a distinction between delete-because-of-rename and other delete operations, and we won't be hydrating in the non-rename case anyway. For now, perhaps change everything to `CallbackPolicy_UserInitiatedOnly` and put a TODO: comment anywhere we might want to have an `AllowAny` policy in future.

To be honest I think the whole function needs a complete overhaul as I've said in #900. My current favoured approach would be not to interleave decision logic with actually doing anything destructive, i.e. sending the messages. So collect the messages we want to send, and once we've made it all the way through to decide what we want to do, actually go and send them. Separating those 2 aspects will also make it easier to only gather necessary information (such as the virtualisation root handle) once rather than keep going back for the same thing over and over.

Reply:
Thanks @pmj, that does make sense!

Do we think we'll use `CallbackPolicy_AllowOnly` in the future?  It looks like it was only on deletes.  Should system processes be deleting files?

Reply:
> Do we think we'll use `CallbackPolicy_AllowOnly` in the future? It looks like it was only on deletes. Should system processes be deleting files?

In some cases, user-initiated file operations in Finder will be farmed out to a root-owned helper tool. This has `launchd` as its parent process, so `CallbackPolicy_UserInitiatedOnly` would block it, for example.

Reply:
**TLDR:** I was wrong, we will not always have looked up the root by now.  When deleting a file we only look up the root if we're hydrating the file first.

----

**Details**

Taking a closer look, I'm not sure the old code was properly handling `UserInitiatedOnly` vs `AllowAny` in the first place for *directories*.

The old logic did not bail out after sending the PreDelete unless either:
- `TryGetVirtualizationRoot` failed
- `ProviderMessaging_TrySendRequestAndWaitForResponse` returned false

As best I can tell, this means that if a system process was trying to delete a *directory*, the following would occur:
1. PreDelete notification is sent to the provider
2. The root lookup here fails due to the policy restriction:

```
// Prevent system services from expanding directories as part of enumeration as this tends to cause deadlocks with the kauth listeners for Antivirus software
if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, CallbackPolicy_UserInitiatedOnly, &root, &vnodeFsidInode, &kauthResult, kauthError))	                
{
    goto CleanupAndReturn;
}
```

End result: The delete is denied, and so there was no point in sending a PreDelete message in the first place.

I think that leaves us with the file scenario, it which case I was wrong.   We might have to look up the root again because we would have only looked up the root if the file was empty and needed to be hydrated.

So I think we could simplify this code to always use `CallbackPolicy_AllowAny` (we would have bailed out early if something required ` CallbackPolicy_UserInitiatedOnly` and that lookup failed).  Additionally (and ideally) we would *only* make this call if we didn't already look up the root somewhere above.

@pmj @jeschu1 does this sound right to you?

Reply:
**Looking at the old code we did:**
`        ProviderCallbackPolicy callbackPolicy = isDirectory ? CallbackPolicy_UserInitiatedOnly : CallbackPolicy_AllowAny;`

**Previous Behavior was**
System initiated directory `DELETE` sends no messages
System initiated file  `DELETE` sends `MessageType_KtoU_NotifyFilePreDelete` message but not  `MessageType_KtoU_HydrateFile`.

**The way the code is structured now**
System initiated directory `DELETE` sends no messages
System initiated file  `DELETE` on a hydrated file, we will always send `MessageType_KtoU_NotifyFilePreDelete`.
System initiated file  `DELETE` on an unhydrated file, sends no messages

Since we now remove the placeholder tag on `MessageType_KtoU_NotifyFilePreDelete` the file must be hydrated before calling it.

@pmj is the desired functionality that we allow  `DELETE` to process regardless of who initiated them?  If so I can change make those changes accordingly so we'll both hydrate and call predelete.  Would `DELETE` be the only command we'd want to ignore owner for at this time?

Reply:
The problem with system processes is primarily hydration/expansion, because that causes file writes by the provider which in turn causes callbacks in e.g. an AV kext's kauth handler, which causes activity in the AV daemon, etc.

So I think we basically want `CallbackPolicy_AllowAny` for any code path that doesn't cause hydration/expansion.

Reply:
I think the simplest approach for now is to do the `CallbackPolicy_AllowAny` as suggested.  Thanks @pmj and @wilbaker!  If we reach this point in code the file must be hydrated so it's safe to make this change.

@pmj a new message type is definitely an option, but considering we're reworking renames I wouldn't do the change in this PR as it's a bit more invasive.

I'd like to discuss callbackPolicy more offline, but definitely not a blocker or change for this PR.